### PR TITLE
DB engine optimize RAM usage

### LINKED ISSUE #6010: Optimize the memory footprint of the Database Engine
Optimize the memory footprint of the Database Engine of #5303.
6% of the uncompressed data is high and should be improved.

### PR DESCRIPTION
**Pull request by mfundul**
##### Summary
Fixes #6010
##### Component Name
database/engine
##### Additional Information
The memory requirements of the new `dbengine` mode are too high. This PR drops them to about half.

### CODE REVIEW

File: daemon/global_statistics.c [hunk 1]
```diff
@@ -535,10 +535,10 @@ void global_statistics_charts(void) {
 
 #ifdef ENABLE_DBENGINE
     if (localhost->rrd_memory_mode == RRD_MEMORY_MODE_DBENGINE) {
-        unsigned long long stats_array[27];
+        unsigned long long stats_array[28];
```
Comment (paulkatsoulakis) [inline daemon/global_statistics.c]: Please kill the magic numbers

File: database/engine/rrdengineapi.c [hunk 2]
```diff
@@ -361,11 +367,12 @@ void *rrdeng_get_page(struct rrdengine_instance *ctx, uuid_t *id, usec_t point_i
...
     }
     *handle = descr;
+    pg_cache_descr = descr->pg_cache_descr;
 
-    return descr->page;
+    return pg_cache_descr->page;
 }
 
-void rrdeng_get_27_statistics(struct rrdengine_instance *ctx, unsigned long long *array)
+void rrdeng_get_28_statistics(struct rrdengine_instance *ctx, unsigned long long *array)
```
Comment (paulkatsoulakis) [inline database/engine/rrdengineapi.c]: Why not `rrdeng_get_statistics()` and just use proper definitions for the indexes below? (last entry could be RRD_ENG_STATISTICS_MAX or something)

File: database/engine/rrdenginelib.c [hunk 3]
```diff
@@ -1,25 +1,49 @@
 // SPDX-License-Identifier: GPL-3.0-or-later
 #include "rrdengine.h"
 
-void print_page_cache_descr(struct rrdeng_page_cache_descr *page_cache_descr)
+/* Caller must hold descriptor lock */
+void print_page_cache_descr(struct rrdeng_page_descr *descr)
 {
+    struct page_cache_descr *pg_cache_descr = descr->pg_cache_descr;
     char uuid_str[37];
     char str[512];
```
Comment (paulkatsoulakis) [inline database/engine/rrdenginelib.c]: More magic numbers, i see a printf below that has that 512 in order to fill str

File: database/engine/rrdenginelib.c [hunk 4]
```diff
@@ -1,25 +1,49 @@
...
     }
-    snprintfz(str + pos, 512 - pos, " flags:0x%2.2lX refcnt:%u\n\n", page_cache_descr->flags, page_cache_descr->refcnt);
+    snprintfz(str + pos, 512 - pos, " flags:0x%2.2lX refcnt:%u\n\n", pg_cache_descr->flags, pg_cache_descr->refcnt);
+    fputs(str, stderr);
+}
+
+void print_page_descr(struct rrdeng_page_descr *descr)
+{
+    char uuid_str[37];
+    char str[512];
```
Comment (paulkatsoulakis) [inline database/engine/rrdenginelib.c]: ditto

File: database/engine/rrdenglocking.c [hunk 5]
```diff
@@ -0,0 +1,209 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+#include "rrdengine.h"
+
+struct page_cache_descr *rrdeng_create_pg_cache_descr(struct rrdengine_instance *ctx)
+{
+    struct page_cache_descr *pg_cache_descr;
+
+    pg_cache_descr = mallocz(sizeof(*pg_cache_descr));
```
Comment (paulkatsoulakis) [inline database/engine/rrdenglocking.c]: how about callocz() and ditch the rest of the lines of manual assignment?

File: database/engine/rrdenglocking.c [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+                /* success */
+                break;
+            }
+            continue; /* spin */
+        }
+        if (old_state & PG_CACHE_DESCR_LOCKED) {
+            assert(0 == old_users);
+            continue; /* spin */
+        }
+        if (0 == old_state) {
```
Comment (paulkatsoulakis) [inline database/engine/rrdenglocking.c]: magic numbers, unless this is indeed some counter

File: database/engine/rrdengineapi.c (hunk 2 above)
Comment (mfundul) [reply database/engine/rrdengineapi.c]: I want it to break whenever somebody decides to add a new statistic cause the risk is great.

File: database/engine/rrdenginelib.c (hunk 3 above)
Comment (mfundul) [reply database/engine/rrdenginelib.c]: Yeah I should add this 512 to a define.

File: database/engine/rrdenginelib.c (hunk 4 above)
Comment (mfundul) [reply database/engine/rrdenginelib.c]: yeap

File: database/engine/rrdenglocking.c (hunk 6 above)
Comment (mfundul) [reply database/engine/rrdenglocking.c]: That's no magic number nor a counter. It's a complex state word described in `pagecache.h`.

**Comment by paulkatsoulakis** [comment]
you can ignore the secondary travis that is showing up pending.
It's a glitch from a configuration @cakrit was trying

**Pull request merged**